batch/batch/driver/scheduler.py

```diff
@@ -111,33 +120,143 @@ def allocate_cores(user, mark):
         async_work = []
-        async for user_record in user_records:
-            records = self.db.execute_and_fetchall(
-                '''
-SELECT attempts.job_id, attempts.batch_id, attempts.attempt_id, instance_name
-FROM attempts
-STRAIGHT_JOIN batches ON batches.id = attempts.batch_id
-STRAIGHT_JOIN jobs ON jobs.batch_id = attempts.batch_id AND jobs.job_id = attempts.job_id
-WHERE jobs.state = 'Running' AND (NOT jobs.always_run) AND batches.closed AND batches.cancelled AND batches.user = %s
-LIMIT 50;
+        for user, share in user_share.items():
+            remaining = Box(share)
+            async for record in user_cancelled_ready_jobs(user, remaining):
+                batch_id = record['batch_id']
+                job_id = record['job_id']
+                id = (batch_id, job_id)
+                log.info(f'cancelling job {id}')
+
+                async def cancel_with_error_handling(id, f):
+                    try:
+                        await f()
+                    except Exception:
+                        log.info(f'error while cancelling job {id}', exc_info=True)
+                async_work.append(
+                    functools.partial(
+                        cancel_with_error_handling,
+                        id,
+                        functools.partial(mark_job_complete,
+                                          self.app, batch_id, job_id, None, None,
+                                          'Cancelled', None, None, None, 'cancelled')))
+
+                remaining.value -= 1
+                if remaining.value <= 0:
+                    should_wait = False
+                    break
+
+        await bounded_gather(*[x for x in async_work], parallelism=100)
+        return should_wait
+
+    async def cancel_cancelled_running_jobs_loop_body(self):
+        records = self.db.select_and_fetchall(
+            '''
+SELECT user, n_cancelled_running_jobs
+FROM (SELECT user,
+    CAST(COALESCE(SUM(n_cancelled_running_jobs), 0) AS SIGNED) AS n_cancelled_running_jobs
+  FROM user_resources
+  GROUP BY user) AS t
+WHERE n_cancelled_running_jobs > 0;
+''')
+        user_n_cancelled_running_jobs = {
+            record['user']: record['n_cancelled_running_jobs'] async for record in records
+        }
+
+        total = sum(user_n_cancelled_running_jobs.values())
+        user_share = {
+            user: max(int(1000 * user_n_jobs / total + 0.5), 20)
+            for user, user_n_jobs in user_n_cancelled_running_jobs.items()
+        }
+
+        async def user_cancelled_running_jobs(user, remaining):
+            async for batch in self.db.select_and_fetchall(
+                    '''
+SELECT id
+FROM batches
+WHERE user = %s AND `state` = 'running' AND cancelled = 1;
+''',
+                    (user,)):
+                async for record in self.db.select_and_fetchall(
+                        '''
+SELECT jobs.job_id, jobs.attempt_id, attempts.instance_name
+FROM jobs
+STRAIGHT_JOIN attempts
```

Comment:
As things stand, I don't know how to write query that way because there is no suitable index on attempts.  That won't be efficient.

I changed it to not join on attempt_id, so we get every attempt for running, cancelled jobs.  We will not get "orphaned" attempts that are running for jobs we think are ready.  I think that's something we can index, let's revisit it another time.